packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -451,15 +472,100 @@ export class DbTargetGlobal {
      * until the backend has caught up to this snapshot version again. This
      * prevents our cache from ever going backwards in time.
      */
-    public lastRemoteSnapshotVersion: DbTimestamp
+    public lastRemoteSnapshotVersion: DbTimestamp,
+    /**
+     * A cache of the number of targets persisted.
+     */
+    public targetCount: number
   ) {}
 }
 
+function createQueryCache(db: IDBDatabase): void {
+  const targetDocumentsStore = db.createObjectStore(DbTargetDocument.store, {
+    keyPath: DbTargetDocument.keyPath as KeyPath
+  });
+  targetDocumentsStore.createIndex(
+    DbTargetDocument.documentTargetsIndex,
+    DbTargetDocument.documentTargetsKeyPath,
+    { unique: true }
+  );
+
+  const targetStore = db.createObjectStore(DbTarget.store, {
+    keyPath: DbTarget.keyPath
+  });
+
+  // NOTE: This is unique only because the TargetId is the suffix.
+  targetStore.createIndex(
+    DbTarget.queryTargetsIndexName,
+    DbTarget.queryTargetsKeyPath,
+    { unique: true }
+  );
+  db.createObjectStore(DbTargetGlobal.store);
+}
+
 /**
- * The list of all IndexedDB stored used by the SDK. This is used when creating
- * transactions so that access across all stores is done atomically.
+ * Counts the number of targets persisted and adds that value to the target
+ * global singleton.
+ * @param {PersistencePromise<void>} p The promise used to keep the transaction open
+ * @param {IDBTransaction} txn The version upgrade transaction for indexeddb
+ * @return {PersistencePromise<void>} A promise of the operations to add the count to the metadata row.
  */
-export const ALL_STORES = [
+function addTargetCount(
+  p: PersistencePromise<void>,
+  txn: IDBTransaction
+): PersistencePromise<void> {
+  const globalStore = txn.objectStore(DbTargetGlobal.store);
+  return p
+    .next(() => wrapRequest<number>(txn.objectStore(DbTarget.store).count()))
+    .next((count: number) =>
+      wrapRequest<DbTargetGlobal>(globalStore.get(DbTargetGlobal.key)).next(
+        (metadata: DbTargetGlobal) => {
+          metadata.targetCount = count;
+          return metadata;
+        }
+      )
+    )
+    .next((metadata: DbTargetGlobal) =>
+      wrapRequest<void>(globalStore.put(metadata, DbTargetGlobal.key))
+    );
+}
+
+/**
+ * Ensures that the target global singleton row exists by adding it if it's missing.
+ * @param {PersistencePromise<void>} p The promise used to keep the transaction open
+ * @param {IDBTransaction} txn The version upgrade transaction for indexeddb
+ * @return {PersistencePromise<void>} A promise of operations to ensure that the metadata row exists
+ */
+function ensureTargetGlobal(
```

Comment:
This function could also return a targetGlobal, which you could then use in addTargetCount.

Reply:
I don't feel super strongly about this, but I do have a slight preference for keeping schema changes independent. Even though these two are in the same version, the actions aren't really strongly related, and the reason for the return value would be convenience.

Reply:
Ok, @mikelehen asked for this change too, so I guess I'm outvoted.